js_modules/dagster-ui/packages/ui-core/src/performance/TraceContext.tsx

```diff
@@ -51,31 +53,35 @@ export class Dependency {
 }
 
 /** Use this to declare a dependency on an apollo query result */
-export function useBlockTraceOnQueryResult(queryResult: QueryResult<any>, name: string) {
-  const dep = useDependency(name);
+export function useBlockTraceOnQueryResult(queryResult: QueryResult<any, any>, name: string) {
+  const dep = useTraceDependency(name);
   const hasData = !!queryResult.data;
   const hasError = !!queryResult.error;
 
-  useLayoutEffect(() => {
+  useDangerousRenderEffect(() => {
     if (hasData) {
       dep.completeDependency(CompletionType.SUCCESS);
     }
   }, [dep, hasData]);
 
-  useLayoutEffect(() => {
+  useDangerousRenderEffect(() => {
     if (!hasData && hasError) {
       dep.completeDependency(CompletionType.ERROR);
     }
   }, [dep, hasData, hasError]);
 }
 
-export function useDependency(name: string) {
+export function useTraceDependency(name: string, uid: any = '') {
   const {addDependency, completeDependency, createDependency} = useContext(TraceContext);
 
-  const dependency = useMemo(() => createDependency(name), [createDependency, name]);
+  // eslint-disable-next-line react-hooks/exhaustive-deps
+  const dependency = useMemo(() => createDependency(name), [createDependency, name, uid]);
```

Comment:
Just want to verify - createDependency includes some randomness? It seems like we use the uid to refresh this value (and need a new value) but the uid is not actually part of the createDependency call.

Reply:
No randomness.

Look at the usage in asyncGraphLayout. It's being used to create a new dependency. The main reason for doing it like this was dev ergonomics. Otherwise you need to manage the dependency yourself.